pkg/verify/verify.go

```diff
@@ -0,0 +1,203 @@
+		}
+		var hashes [][]byte
+		for _, h := range consistencyProof.Payload.Hashes {
+			b, err := hex.DecodeString(h)
+			if err != nil {
+				return errors.New("error decoding consistency proof hashes")
+			}
+			hashes = append(hashes, b)
+		}
+		if err := proof.VerifyConsistency(rfc6962.DefaultHasher,
+			uint64(oldTreeSize), sth.Size, hashes, oldStateHash, sth.Hash); err != nil {
+			return err
+		}
+	case oldTreeSize > int64(sth.Size):
+		return errors.New("inclusion proof returned a tree size larger than the verified tree size")
+	}
+	return nil
+
+}
+
+// VerifyRootHash verifies the provided root hash at the treeSize by verifying consistenct
+// against an STH.
+//nolint
+func VerifyRootHash(ctx context.Context, rClient *client.Rekor, verifier signature.Verifier,
+	rootHash []byte, treeSize int64, treeID string) error {
+	// Get and verify the current STH.
+	infoParams := tlog.NewGetLogInfoParamsWithContext(ctx)
+	result, err := rClient.Tlog.GetLogInfo(infoParams)
+	if err != nil {
+		return err
+	}
+
+	logInfo := result.GetPayload()
+
+	sth := util.SignedCheckpoint{}
+	if err := sth.UnmarshalText([]byte(*logInfo.SignedTreeHead)); err != nil {
+		return err
+	}
+
+	if !sth.Verify(verifier) {
+		return errors.New("signature on tree head did not verify")
+	}
+
+	// Now verify consistency up to the STH.
+	return ProveConsistency(ctx, rClient, rootHash, treeSize, sth, treeID)
+}
+
+// VerifyInclusion verifies an entry's inclusion proof. Clients MUST also
+// verify the root hash via VerifyRootHash.
+//nolint
+func VerifyInclusion(ctx context.Context, e *models.LogEntryAnon) error {
+	if e.Verification == nil || e.Verification.InclusionProof == nil {
+		return errors.New("inclusion proof not provided")
+	}
+
+	hashes := [][]byte{}
+	for _, h := range e.Verification.InclusionProof.Hashes {
+		hb, _ := hex.DecodeString(h)
+		hashes = append(hashes, hb)
+	}
+
+	rootHash, err := hex.DecodeString(*e.Verification.InclusionProof.RootHash)
+	if err != nil {
+		return err
+	}
+
+	// Verify the inclusion proof.
+	entryBytes, err := base64.StdEncoding.DecodeString(e.Body.(string))
+	if err != nil {
+		return err
+	}
+	leafHash := rfc6962.DefaultHasher.HashLeaf(entryBytes)
+
+	if err := proof.VerifyInclusion(rfc6962.DefaultHasher, uint64(*e.Verification.InclusionProof.LogIndex),
+		uint64(*e.Verification.InclusionProof.TreeSize), leafHash, hashes, rootHash); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// VerifySignedEntryTimestamp verifies the entry's SET against the provided
+// public key.
+//nolint
+func VerifySignedEntryTimestamp(ctx context.Context, e *models.LogEntryAnon, verifier signature.Verifier) error {
+	if e.Verification == nil {
+		return fmt.Errorf("missing verification")
+	}
+	if e.Verification.SignedEntryTimestamp == nil {
+		return fmt.Errorf("signature missing")
+	}
+
+	type bundle struct {
+		Body           interface{} `json:"body"`
+		IntegratedTime int64       `json:"integratedTime"`
+		// Note that this is the virtual index.
```

Comment:
I thought we were supposed to verify against the shard log index, not the virtual log index?

Reply:
Nope. Inclusion proof requires verification against the internal log index. But we sign the bundle over the virtual index:
https://github.com/sigstore/rekor/blob/6093a3391c2399d384b50bb9b6f48e4011111729/pkg/api/entries.go#L214-L258

Reply:
gotcha, thanks!

Reply:
i don't think it's a security issue either way though -- since the signed payload also contains the log ID, so it disambiguates between shards anyway